runtime/evaluate.cc

```diff
@@ -173,7 +173,11 @@ class evaluator : public node_visitor {
 
   template <typename T>
   void visit_let(const T* op) {
-    auto set_value = set_value_in_scope(context, op->sym, eval_expr(op->value));
+    std::vector<scoped_value_in_symbol_map<index_t>> scoped_values;
```

Comment:
We should try to do this on the stack here. I'm trying to avoid any heap allocations in `evaluate` (except for explicitly requested buffer allocations on the heap).

I think the best thing to do is just `alloca` `std::optional<index_t>` x `op->lets.size()`, and just set the values in the map manually (I think using the `set_value_in_scope` helper is going to make it harder, not easier).

Reply:
See the `loop` visitor (the else branch for `loop_mode::serial`) for an example of how to do this.

Reply:
using `alloca` for a nontrivial type like `std::optional` is tricky since the ctors/dtors don't automatically run. I can make the logic work using placement new and explicit calls to dtors but it will be a bit ugly.